python/prairielearn.py

```diff
@@ -56,6 +68,105 @@ class ElementTestData(QuestionData):
+    weight: int = 1,
+) -> None:
+    """
+    Grade question question_name. grade_function should take in a single parameter
+    (which will be the submitted answer) and return a 2-tuple:
+        - The first element of the 2-tuple should either be:
+            - a boolean indicating whether the question should be marked correct
+            - a partial score between 0 and 1, inclusive
+        - The second element of the 2-tuple should either be:
+            - a string containing feedback
+            - None, if there is no feedback (usually this should only occur if the answer is correct)
+    """
+
+    # Create the data dictionary at first
+    data["partial_scores"][question_name] = {"score": 0.0, "weight": weight}
+
+    if question_name not in data["submitted_answers"]:
+        data["format_errors"][question_name] = "No answer was submitted"
+        return
+
+    submitted_answer = data["submitted_answers"][question_name]
+
+    # Run passed-in grading function
+    result, feedback_content = grade_function(submitted_answer)
+
+    # Try converting partial score
+    if isinstance(result, bool):
+        partial_score = 1.0 if result else 0.0
+    elif isinstance(result, (float, int)):
+        assert 0.0 <= result <= 1.0
+        partial_score = result
+    else:
+        assert_never(result)
+
+    # Set corresponding partial score and feedback
+    data["partial_scores"][question_name]["score"] = partial_score
+
+    if feedback_content:
+        data["partial_scores"][question_name]["feedback"] = feedback_content
+
+
+def determine_score_params(score: float) -> Tuple[str, Union[bool, float]]:
+    """Determine score params taken from data dict"""
+
+    if score >= 1:
+        return ("correct", True)
+    elif score > 0:
+        return ("partial", math.floor(score * 100))
+
+    return ("incorrect", True)
+
+
+EnumT = TypeVar("EnumT", bound=Enum)
+
+
+def get_enum_attrib(
+    element: lxml.html.HtmlElement,
+    name: str,
+    enum_type: Type[EnumT],
+    default: Optional[EnumT] = None,
+) -> EnumT:
+    """
+    Returns the named attribute for the element parsed as an enum,
+    or the (optional) default value. If the default value is not provided
+    and the attribute is missing then an exception is thrown. An exception
+    is also thrown if the value for the enum provided is invalid.
+    Also, alters the enum names to comply with PL naming convention automatically
+    (replacing underscores with dashes and uppercasing). If default value is
+    provided, must be a member of the given enum.
+    """
+
+    enum_val, is_default = (
+        _get_attrib(element, name)
+        if default is None
+        else _get_attrib(element, name, default)
+    )
+
+    # Default doesn't need to be converted, already a value of the enum
+    if is_default:
+        return enum_val
+
+    upper_enum_str = enum_val.upper()
+    accepted_names = {member.name.replace("_", "-") for member in enum_type}
```

Comment:
do we really need/want to accept the underscored version of the name here? I know this is down with the other ones for backwards compatibility, but we shouldn't need backwards compatibility since this is new functionality.

Reply:
ok cool